docs/create_your_own_brick.rst

```diff
@@ -0,0 +1,387 @@
+advocated whenever it makes sense.
+
+Here are examples of possible bricks to inherit from:
+
+* :class:`.Sequence`: a sequence of bricks.
+* :class:`.Initializable`: a brick that defines a same initialiation scheme
+  (weights and biases) for all its children.
+* :class:`.Feedforward`: declares an interface for bricks with one input and
+  one output.
+* :class:`.Linear`: a linear transformation with optional bias. Inherits from
+  :class:`.Initializable` and :class:`.Feedforward`.
+* :class:`.BaseRecurrent`: the base class for recurrent bricks. Check the
+  :doc:`tutorial about rnns</rnn>` for more information.
+* many more!
+
+Let's say that you want to create a brick from scracth, simply inheriting
+from :class:`.Brick`, then you should consider overwriting the following
+methods (strictly speaking, all these methods are optional):
+
+* :meth:`.Brick.__init__`: you should pass by argument the attributes of your
+  brick. It is also in this method that you should create the potential
+  "children bricks" that belongs to your brick (in that case, you have to put
+  the children bricks into ``self.children``). The initialiazation of the
+  attributes can be lazy as described later in the tutorial.
+* :meth:`apply`: you need to implement a method that actually
+  implements the operation of the brick, taking as arguments the inputs
+  of the brick and returning its outputs. It can have any name and for simple
+  bricks is often named ``apply``. You should decorate it with the
+  :func:`.application` decorator, as explained in the next section. If you
+  design a recurrent brick, you should instead decorate it with the
+  :func:`.recurrent` decorator as explained in the
+  :doc:`tutorial about rnns</rnn>`.
+* :meth:`.Brick._allocate`: you should implement this method to allocate the
+  shared variables (often representing parameters) of the brick. In blocks,
+  by convention, the built-in bricks allocate their shared variables with nan
+  values and we recommend you to do the same.
+* :meth:`.Brick._initialize`: you should implement this method to initialize
+  the shared variables of your brick. This method is called after the
+  allocation.
+* :meth:`.Brick._push_allocation_config`: you should consider overwriting
+  this method if you want to change configuration of the children bricks
+  before they allocate their parameters.
+* :meth:`.Brick._push_initialization_config`: you should consider
+  overwriting this method if you want to change the initialization schemes of
+  the children before they get initialized.
+  If the children bricks need to be initialized with the same scheme, then you
+  should inherit your brick from :class:`.Initializable`, which
+  automatically push the initialization schemes of your brick (provided as
+  arguments ``weights_init`` and ``biases_init`` of the constructor) to the
+  children bricks.
+* :meth:`.Brick.get_dim`: implementing this function is useful if you want
```

Comment:
I would like to also explain here why using `get_dim` is preferable to accessing the attributes of the brick. The reason is that `get_dim` is the interface that is the same for all bricks, whereas the exact way allocation configuration is stored in a brick object might vary. For instance, `Feedforward` bricks have `input_dim` and `output_dim` attributes, whereas `Parallel` bricks have `input_dims` and `output_dims` (note plural instead of singular). When one wants to change the configuration of a brick, knowing how the attributes are called and what they mean is necessary, but when only read-only access required, `get_dim` should be sufficient. Thus, it was introduced to facilitate writing bricks that use other bricks without making too many assumptions about their interface.

Please let me know if this explanation was helpful and if you can integrate it in a concise way into your tutorial.